lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -263,6 +270,64 @@ deriveAddressPrivateKey (Passphrase pwd) accountKey idx@(Index addrIx) = RndKey
     , payloadPassphrase = payloadPassphrase accountKey
     }
 
+{-------------------------------------------------------------------------------
+                          Storing and retrieving keys
+-------------------------------------------------------------------------------}
+
+instance PersistKey RndKey where
+    serializeXPrv = serializeXPrvRnd
+    deserializeXPrv = deserializeXPrvRnd
+    deserializeXPub = deserializeXPubRnd
+    serializeXPub = serializeXPubRnd
+
+-- | Serialize the key with its payload encryption passphrase.
+serializeKey :: (ByteString, Passphrase "addr-derivation-payload") -> ByteString
+serializeKey (k, Passphrase p) =
+    convertToBase Base16 k <> ":" <> convertToBase Base16 p
+
+-- | Deserialize the key and its payload encryption passphrase.
+deserializeKey
+    :: (ByteString -> Either String key)
+    -> ByteString
+    -> Either String (key, Passphrase "addr-derivation-payload")
+deserializeKey f b = case map (convertFromBase Base16) (B8.split ':' b) of
+    [Right k, Right p] -> case f k of
+        Right k' -> Right (k', Passphrase (BA.convert p))
+        Left e -> Left e
+    _ -> Left "Key input must be two hex strings separated by :"
+
+serializeXPrvRnd
+    :: (RndKey depth XPrv, Hash "encryption")
+    -> (ByteString, ByteString)
+serializeXPrvRnd (RndKey k _ p, h) =
+    ( serializeKey (unXPrv k, p)
+    , convertToBase Base16 . getHash $ h )
+
+-- | The reverse of 'serializeXPrv'. This may fail if the inputs are not valid
+-- hexadecimal strings, or if the key is of the wrong length.
+deserializeXPrvRnd
+    :: (ByteString, ByteString)
+       -- ^ Hexadecimal encoded private key and password hash
+    -> Either String (RndKey 'RootK XPrv, Hash "encryption")
+deserializeXPrvRnd (k, h) = (,)
+    <$> fmap mkKey (rootKeyFromText k)
+    <*> fmap Hash (convertFromBase Base16 h)
+  where
+    rootKeyFromText = deserializeKey xprv
+    mkKey (key, pwd) = RndKey key () pwd
+
+-- | Convert a public key into a hexadecimal string suitable for storing in
+-- a text file or database column.
+serializeXPubRnd :: RndKey depth XPub -> ByteString
+serializeXPubRnd (RndKey k _ p) = serializeKey (unXPub k, p)
+
+-- | The reverse of 'serializeXPub'. This will fail if the input is not a valid
+-- hexadecimal string of the correct length.
+deserializeXPubRnd :: ByteString -> Either String (RndKey 'RootK XPub)
+deserializeXPubRnd = fmap mkKey . (deserializeKey xpub)
```

Comment:
Note that I think we only need to serialize public key in order to serialize `AddressPool` in the sequential derivation. I don't think we have to serialize / store any random public keys actually :thinking: ?

Reply:
I should have checked where it was used. That is right, public keys only need to be serialized for SeqState. Fixed.